server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -365,31 +379,70 @@ private void snapshot(
             }
 
             final Repository repository = repositoriesService.repository(snapshot.getRepository());
-            Engine.IndexCommitRef snapshotRef = null;
+            SnapshotIndexCommit snapshotIndexCommit = null;
             try {
-                snapshotRef = indexShard.acquireIndexCommitForSnapshot();
+                snapshotIndexCommit = new SnapshotIndexCommit(indexShard.acquireIndexCommitForSnapshot());
+                snapshotStatus.addAbortListener(makeAbortListener(indexShard.shardId(), snapshot, snapshotIndexCommit));
                 snapshotStatus.ensureNotAborted();
                 repository.snapshotShard(
                     new SnapshotShardContext(
                         indexShard.store(),
                         indexShard.mapperService(),
                         snapshot.getSnapshotId(),
                         indexId,
-                        new SnapshotIndexCommit(snapshotRef),
-                        getShardStateId(indexShard, snapshotRef.getIndexCommit()),
+                        snapshotIndexCommit,
+                        getShardStateId(indexShard, snapshotIndexCommit.indexCommit()),
                         snapshotStatus,
                         version,
                         entryStartTime,
                         listener
                     )
                 );
-            } catch (Exception e) {
-                IOUtils.close(snapshotRef);
-                throw e;
+                snapshotIndexCommit = null; // success
+            } finally {
+                if (snapshotIndexCommit != null) {
+                    try {
+                        snapshotIndexCommit.onCompletion();
+                    } catch (Exception e) {
+                        // we're already failing exceptionally, and prefer to propagate the original exception instead of this one
+                        logger.warn(Strings.format("exception when closing commit for [%s] in [%s]", indexShard.shardId(), snapshot), e);
+                    }
+                }
             }
         });
     }
 
+    private static ActionListener<IndexShardSnapshotStatus.AbortStatus> makeAbortListener(
+        ShardId shardId,
+        Snapshot snapshot,
+        SnapshotIndexCommit snapshotIndexCommit
+    ) {
+        return new ActionListener<>() {
+            @Override
+            public void onResponse(IndexShardSnapshotStatus.AbortStatus abortStatus) {
+                if (abortStatus == IndexShardSnapshotStatus.AbortStatus.ABORTED) {
+                    assert ThreadPool.assertCurrentThreadPool(ThreadPool.Names.GENERIC, ThreadPool.Names.SNAPSHOT);
+                    snapshotIndexCommit.onAbort();
```

Comment:
It seems there could be a race condition involved around `closeException` in `SnapshotIndexCommit` in case `onAbort` and `onComplete` runs concurrently (as you pointed out, `decRef` might not have run before `onComplete` checks the `closeException`), the happens-before mentioned [here](https://github.com/elastic/elasticsearch/pull/96426/files#diff-35fe592a1471f816a2bea7eb7075fd0fa2a217cc31e8931055fd9a0cd81947efR50) is towards the `getAndSet` in `RunOnce`, where the `decRef` happens after that.

Reply:
:+1: yes sorry for missing this earlier. #96571 will fix this up.